csrc/fusion_profiler.h

```diff
@@ -0,0 +1,319 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <chrono>
+#include <unordered_map>
+
+#include <c10/cuda/CUDAStream.h>
+#include <cuda_runtime.h>
+#include <cuda_utils.h>
+#include <cupti.h>
+#include <debug.h>
+#include <options.h>
+#include <utils.h>
+
+namespace nvfuser {
+
+//! \enum ProfilerState
+//! \brief An enum used to represent the state of a profiling state machine
+enum class ProfilerState {
+  Ready,
+  Running,
+  Finished,
+  Processed,
+};
+
+std::ostream& operator<<(std::ostream&, const ProfilerState&);
+
+//! \class CudaEventTimer
+//! \brief A Cuda Events based timer that includes GPU time for kernels launched
+//! between those events.
+class CudaEventTimer {
+ public:
+  CudaEventTimer(cudaStream_t s);
+  ~CudaEventTimer();
+
+  void reset();
+  void start();
+  void stop();
+  double time();
+  ProfilerState state() const;
+
+ private:
+  cudaStream_t stream_;
+  cudaEvent_t start_event_;
+  cudaEvent_t stop_event_;
+  double time_ms_;
+  ProfilerState state_;
+};
+
+//! \class HostTimer
+//! \brief A std::chrono::stead_clock based timer of CPU activity.
+class HostTimer {
+ public:
+  using Clock = std::chrono::steady_clock;
+
+  HostTimer();
+
+  void reset();
+  void start();
+  void stop();
+  double time();
+  ProfilerState state() const;
+
+ private:
+  Clock::time_point start_event_;
+  Clock::time_point stop_event_;
+  double time_ms_;
+  ProfilerState state_;
+};
+
+//! \struct DeviceDescriptor
+//! \brief This struct captures the GPU information necessary to calculate the
+//! the Peak Bandwidth of the specific GPU queried.
+struct DeviceDescriptor {
+  //! Queries the GPU to populate the struct's data members and calculates the
+  //! peak bandwidth
+  void generate(int device);
```

Comment:
`generate` doesn't need to be an instance method. I'd say

```
static DeviceDescriptor generate(int device);
```

Reply:
I made this change.